examples/model-viewer/main.cpp

```diff
@@ -0,0 +1,922 @@
+
+        shaderProgram = device->createProgram(programDesc);
+
+        // Get other shader types that we will use for creating shader objects.
+        perViewShaderType = slangReflection->findTypeByName("PerView");
+        perModelShaderType = slangReflection->findTypeByName("PerModel");
+
+        return SLANG_OK;
+    }
+};
+
+// Our application code has a rudimentary material system,
+// to match the `IMaterial` abstraction used in the shade code.
+//
+struct Material : RefObject
+{
+    // The key feature of a matrial in our application is that
+    // it can provide a shader object that describes it and
+    // its parameters. The contents of the shader object will
+    // be any colors, textures, etc. that the material needs,
+    // while the Slang type that was used to allocate the
+    // block will be an implementation of `IMaterial` that
+    // provides the evaluation logic for the material.
+
+    // Each subclass of `Material` will provide a routine to
+    // create a shader object that stores its shader parameters.
+    virtual IShaderObject* createShaderObject(RendererContext* context) = 0;
+
+    // The shader object for a material will be stashed here
+    // after it is created.
+    ComPtr<IShaderObject> shaderObject;
+};
+
+// For now we have only a single implementation of `Material`,
+// which corresponds to the `SimpleMaterial` type in our shader
+// code.
+//
+struct SimpleMaterial : Material
+{
+    glm::vec3   diffuseColor;
+    glm::vec3   specularColor;
+    float       specularity;
+
+    // Create a shader object that contains the type info and parameter values
+    // that represent an instance of `SimpleMaterial`.
+    IShaderObject* createShaderObject(RendererContext* context) override
+    {
+        auto program = context->slangReflection;
+        auto shaderType = program->findTypeByName("SimpleMaterial");
+        shaderObject = context->device->createShaderObject(shaderType);
+        gfx::ShaderCursor cursor(shaderObject);
+        cursor["diffuseColor"].setData(&diffuseColor, sizeof(diffuseColor));
+        cursor["specularColor"].setData(&specularColor, sizeof(specularColor));
+        cursor["specularity"].setData(&specularity, sizeof(specularity));
+        return shaderObject.get();
+    }
+};
+
+// With the `Material` abstraction defined, we can go on to define
+// the representation for loaded models that we will use.
+//
+// A `Model` will own vertex/index buffers, along with a list of meshes,
+// while each `Mesh` will own a material and a range of indices.
+// For this example we will be loading models from `.obj` files, but
+// that is just a simple lowest-common-denominator choice.
+//
+struct Mesh : RefObject
+{
+    RefPtr<Material>    material;
+    int                 firstIndex;
+    int                 indexCount;
+};
+struct Model : RefObject
+{
+    typedef platform::ModelLoader::Vertex Vertex;
+
+    ComPtr<IBufferResource>     vertexBuffer;
```

Comment:
Unrelated to this change, but we eventually need to sit down and work out what the ideal system is for making rasterization shaders in Slang be adaptable to work with many vertex layouts (and combinations of vertex attributes). We have solved this with metaprogramming sorts of approaches before, but it seems like we eventually need to demonstrate a more complete solution to the problem.